src/thor/timedep_forward.cc

```diff
@@ -136,57 +145,114 @@ bool TimeDepForward::ExpandForward(GraphReader& graphreader,
   if (meta.edge_status->set() == EdgeSet::kPermanent) {
     return true; // This is an edge we _could_ have expanded, so return true
   }
+
+  graph_tile_ptr t2 = nullptr;
+  GraphId opp_edge_id;
+  const DirectedEdge* opp_edge = nullptr;
+  if (!FORWARD) {
+    // Get end node tile, opposing edge Id, and opposing directed edge.
+    t2 = meta.edge->leaves_tile() ? graphreader.GetGraphTile(meta.edge->endnode()) : tile;
+    if (t2 == nullptr) {
+      return false;
+    }
+    opp_edge_id = t2->GetOpposingEdgeId(meta.edge);
+    opp_edge = t2->directededge(opp_edge_id);
+  }
+
   // Skip shortcut edges for time dependent routes, if no access is allowed to this edge
   // (based on costing method)
   uint8_t restriction_idx = -1;
-  if (meta.edge->is_shortcut() ||
-      !costing_->Allowed(meta.edge, pred, tile, meta.edge_id, time_info.local_time,
-                         nodeinfo->timezone(), restriction_idx) ||
-      costing_->Restricted(meta.edge, pred, edgelabels_, tile, meta.edge_id, true, &edgestatus_,
-                           time_info.local_time, nodeinfo->timezone())) {
-    return false;
+  if (FORWARD) {
+    if (meta.edge->is_shortcut() ||
+        !costing_->Allowed(meta.edge, pred, tile, meta.edge_id, time_info.local_time,
+                           nodeinfo->timezone(), restriction_idx) ||
+        costing_->Restricted(meta.edge, pred, edgelabels_, tile, meta.edge_id, true, &edgestatus_,
+                             time_info.local_time, nodeinfo->timezone())) {
+      return false;
+    }
+  } else {
+    if (!costing_->AllowedReverse(meta.edge, pred, opp_edge, t2, opp_edge_id, time_info.local_time,
+                                  nodeinfo->timezone(), restriction_idx) ||
+        costing_->Restricted(meta.edge, pred, edgelabels_, tile, meta.edge_id, false, &edgestatus_,
+                             time_info.local_time, nodeinfo->timezone())) {
+      return false;
+    }
   }
 
   // Compute the cost to the end of this edge
   uint8_t flow_sources;
-  auto edge_cost = costing_->EdgeCost(meta.edge, tile, time_info.second_of_week, flow_sources);
-  auto transition_cost = costing_->TransitionCost(meta.edge, nodeinfo, pred);
-  Cost newcost = pred.cost() + edge_cost + transition_cost;
+  auto edge_cost = FORWARD
+                       ? costing_->EdgeCost(meta.edge, tile, time_info.second_of_week, flow_sources)
+                       : costing_->EdgeCost(opp_edge, t2, time_info.second_of_week, flow_sources);
+
+  auto transition_cost =
+      FORWARD ? costing_->TransitionCost(meta.edge, nodeinfo, pred)
+              : costing_->TransitionCostReverse(meta.edge->localedgeidx(), nodeinfo, opp_edge,
+                                                opp_pred_edge, pred.has_measured_speed(),
+                                                pred.internal_turn());
+  Cost newcost = pred.cost() + edge_cost;
+  if (FORWARD) {
+    newcost += transition_cost;
+  } else {
+    newcost.cost += transition_cost.cost;
```

Comment:
This was similar to the thing found in the Dijkstra refactor - why does reverse only add cost?  This seems like a bug, as it means a TimeDepReverse path won't accumulate time?